File: common/rust/cctrusted_base/src/tcg.rs
use crate::binary_blob::dump_data;
use hashbrown::HashMap;
use log::info;

pub const TPM_ALG_ERROR: u16 = 0x0;
pub const TPM_ALG_RSA: u16 = 0x1;
pub const TPM_ALG_SHA1: u16 = 0x4;
pub const TPM_ALG_SHA256: u16 = 0xB;
pub const TPM_ALG_SHA384: u16 = 0xC;
pub const TPM_ALG_SHA512: u16 = 0xD;
pub const TPM_ALG_ECDSA: u16 = 0x18;

// hash algorithm ID to algorithm name string map
lazy_static! {
    pub static ref ALGO_NAME_MAP: HashMap<u16, String> = {
        let mut map: HashMap<u16, String> = HashMap::new();
        map.insert(TPM_ALG_ERROR, "TPM_ALG_ERROR".to_string());
        map.insert(TPM_ALG_RSA, "TPM_ALG_RSA".to_string());
        map.insert(TPM_ALG_SHA1, "TPM_ALG_SHA1".to_string());
        map.insert(TPM_ALG_SHA256, "TPM_ALG_SHA256".to_string());
        map.insert(TPM_ALG_SHA384, "TPM_ALG_SHA384".to_string());
        map.insert(TPM_ALG_SHA512, "TPM_ALG_SHA512".to_string());
        map.insert(TPM_ALG_ECDSA, "TPM_ALG_ECDSA".to_string());
        map
    };
}

// this trait retrieve tcg standard algorithm name in string
pub trait TcgAlgorithmRegistry {
    fn get_algorithm_id(&self) -> u16;
    fn get_algorithm_id_str(&self) -> String;
}

// digest format: (algo id, hash value)
#[derive(Clone)]
pub struct TcgDigest {
    pub algo_id: u16,
    pub hash: Vec<u8>,
}

impl TcgDigest {
    pub fn show(&self) {
        info!("show data in struct TcgDigest");
        info!(
            "algo = {}",
            ALGO_NAME_MAP.get(&self.algo_id).unwrap().to_owned()
        );
        info!("hash = {:02X?}", self.hash);
    }

    pub fn get_hash(&self) -> Vec<u8> {
        self.hash.clone()
    }
}

impl TcgAlgorithmRegistry for TcgDigest {
    fn get_algorithm_id(&self) -> u16 {
        self.algo_id
    }

    fn get_algorithm_id_str(&self) -> String {
        ALGO_NAME_MAP.get(&self.algo_id).unwrap().to_owned()
    }
}

// traits a Tcg IMR should have
pub trait TcgIMR {
    fn max_index() -> u8;
    fn get_index(&self) -> u8;
    fn get_tcg_digest(&self, algo_id: u16) -> TcgDigest;
    fn is_valid_index(index: u8) -> Result<bool, anyhow::Error>;
    fn is_valid_algo(algo_id: u16) -> Result<bool, anyhow::Error>;
}

/***
    TCG EventType defined at
    https://trustedcomputinggroup.org/wp-content/uploads/TCG_EFI_Platform_1_22_Final_-v15.pdf
*/
pub const EV_PREBOOT_CERT: u32 = 0x0;
pub const EV_POST_CODE: u32 = 0x1;
pub const EV_UNUSED: u32 = 0x2;
pub const EV_NO_ACTION: u32 = 0x3;
pub const EV_SEPARATOR: u32 = 0x4;
pub const EV_ACTION: u32 = 0x5;
pub const EV_EVENT_TAG: u32 = 0x6;
pub const EV_S_CRTM_CONTENTS: u32 = 0x7;
pub const EV_S_CRTM_VERSION: u32 = 0x8;
pub const EV_CPU_MICROCODE: u32 = 0x9;
pub const EV_PLATFORM_CONFIG_FLAGS: u32 = 0xa;
pub const EV_TABLE_OF_DEVICES: u32 = 0xb;
pub const EV_COMPACT_HASH: u32 = 0xc;
pub const EV_IPL: u32 = 0xd;
pub const EV_IPL_PARTITION_DATA: u32 = 0xe;
pub const EV_NONHOST_CODE: u32 = 0xf;
pub const EV_NONHOST_CONFIG: u32 = 0x10;
pub const EV_NONHOST_INFO: u32 = 0x11;
pub const EV_OMIT_BOOT_DEVICE_EVENTS: u32 = 0x12;

pub const EV_EFI_EVENT_BASE: u32 = 0x80000000;
pub const EV_EFI_VARIABLE_DRIVER_CONFIG: u32 = EV_EFI_EVENT_BASE + 0x1;
pub const EV_EFI_VARIABLE_BOOT: u32 = EV_EFI_EVENT_BASE + 0x2;
pub const EV_EFI_BOOT_SERVICES_APPLICATION: u32 = EV_EFI_EVENT_BASE + 0x3;
pub const EV_EFI_BOOT_SERVICES_DRIVER: u32 = EV_EFI_EVENT_BASE + 0x4;
pub const EV_EFI_RUNTIME_SERVICES_DRIVER: u32 = EV_EFI_EVENT_BASE + 0x5;
pub const EV_EFI_GPT_EVENT: u32 = EV_EFI_EVENT_BASE + 0x6;
pub const EV_EFI_ACTION: u32 = EV_EFI_EVENT_BASE + 0x7;
pub const EV_EFI_PLATFORM_FIRMWARE_BLOB: u32 = EV_EFI_EVENT_BASE + 0x8;
pub const EV_EFI_HANDOFF_TABLES: u32 = EV_EFI_EVENT_BASE + 0x9;
pub const EV_EFI_VARIABLE_AUTHORITY: u32 = EV_EFI_EVENT_BASE + 0x10;

lazy_static! {
    pub static ref TCG_EVENT_TYPE_NAME_MAP: HashMap<u32, String> = {
        let mut map: HashMap<u32, String> = HashMap::new();
        map.insert(EV_PREBOOT_CERT, "EV_PREBOOT_CERT".to_string());
        map.insert(EV_POST_CODE, "EV_POST_CODE".to_string());
        map.insert(EV_UNUSED, "EV_UNUSED".to_string());
        map.insert(EV_NO_ACTION, "EV_NO_ACTION".to_string());
        map.insert(EV_SEPARATOR, "EV_SEPARATOR".to_string());
        map.insert(EV_ACTION, "EV_ACTION".to_string());
        map.insert(EV_EVENT_TAG, "EV_EVENT_TAG".to_string());
        map.insert(EV_S_CRTM_CONTENTS, "EV_S_CRTM_CONTENTS".to_string());
        map.insert(EV_S_CRTM_VERSION, "EV_S_CRTM_VERSION".to_string());
        map.insert(EV_CPU_MICROCODE, "EV_CPU_MICROCODE".to_string());
        map.insert(
            EV_PLATFORM_CONFIG_FLAGS,
            "EV_PLATFORM_CONFIG_FLAGS".to_string(),
        );
        map.insert(EV_TABLE_OF_DEVICES, "EV_TABLE_OF_DEVICES".to_string());
        map.insert(EV_COMPACT_HASH, "EV_COMPACT_HASH".to_string());
        map.insert(EV_IPL, "EV_IPL".to_string());
        map.insert(EV_IPL_PARTITION_DATA, "EV_IPL_PARTITION_DATA".to_string());
        map.insert(EV_NONHOST_CODE, "EV_NONHOST_CODE".to_string());
        map.insert(EV_NONHOST_CONFIG, "EV_NONHOST_CONFIG".to_string());
        map.insert(EV_NONHOST_INFO, "EV_NONHOST_INFO".to_string());
        map.insert(
            EV_OMIT_BOOT_DEVICE_EVENTS,
            "EV_OMIT_BOOT_DEVICE_EVENTS".to_string(),
        );
        map.insert(EV_EFI_EVENT_BASE, "EV_EFI_EVENT_BASE".to_string());
        map.insert(
            EV_EFI_VARIABLE_DRIVER_CONFIG,
            "EV_EFI_VARIABLE_DRIVER_CONFIG".to_string(),
        );
        map.insert(EV_EFI_VARIABLE_BOOT, "EV_EFI_VARIABLE_BOOT".to_string());
        map.insert(
            EV_EFI_BOOT_SERVICES_APPLICATION,
            "EV_EFI_BOOT_SERVICES_APPLICATION".to_string(),
        );
        map.insert(
            EV_EFI_BOOT_SERVICES_DRIVER,
            "EV_EFI_BOOT_SERVICES_DRIVER".to_string(),
        );
        map.insert(
            EV_EFI_RUNTIME_SERVICES_DRIVER,
            "EV_EFI_RUNTIME_SERVICES_DRIVER".to_string(),
        );
        map.insert(EV_EFI_GPT_EVENT, "EV_EFI_GPT_EVENT".to_string());
        map.insert(EV_EFI_ACTION, "EV_EFI_ACTION".to_string());
        map.insert(
            EV_EFI_PLATFORM_FIRMWARE_BLOB,
            "EV_EFI_PLATFORM_FIRMWARE_BLOB".to_string(),
        );
        map.insert(EV_EFI_HANDOFF_TABLES, "EV_EFI_HANDOFF_TABLES".to_string());
        map.insert(
            EV_EFI_VARIABLE_AUTHORITY,
            "EV_EFI_VARIABLE_AUTHORITY".to_string(),
        );
        map
    };
}

#[derive(Clone)]
pub struct TcgEventType {}

impl TcgEventType {
    pub fn get_event_type_string(event_type: u32) -> String {
        match TCG_EVENT_TYPE_NAME_MAP.get(&event_type) {
            Some(str) => str.to_string(),
            None => "UNKNOWN".to_string(),
        }
    }
}

/***
    TCG IMR Event struct defined at
    https://trustedcomputinggroup.org/wp-content/uploads/TCG_EFI_Platform_1_22_Final_-v15.pdf.
    Definition:
    typedef struct tdTCG_PCR_EVENT2{
        UINT32 pcrIndex;
        UINT32 eventType;
        TPML_DIGEST_VALUES digests;
        UINT32 eventSize;
        BYTE event[eventSize];
    } TCG_PCR_EVENT2;
*/
#[derive(Clone)]
pub struct TcgImrEvent {
    pub imr_index: u32,
    pub event_type: u32,
    pub digests: Vec<TcgDigest>,
    pub event_size: u32,
    pub event: Vec<u8>,
}

/***
    TCG TCG_PCClientPCREvent defined at
    https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClientSpecPlat_TPM_2p0_1p04_pub.pdf.
    Definition:
    typedef tdTCG_PCClientPCREvent {
        UINT32 pcrIndex;
        UINT32 eventType;
        BYTE digest[20];
        UINT32 eventDataSize;
        BYTE event[eventDataSize]; //This is actually a TCG_EfiSpecIDEventStruct
    } TCG_PCClientPCREvent;
*/
#[derive(Clone)]
pub struct TcgPcClientImrEvent {
    pub imr_index: u32,
    pub event_type: u32,
    pub digest: [u8; 20],
    pub event_size: u32,
    pub event: Vec<u8>,
}

/***
    TCG TCG_EfiSpecIDEventStruct defined at
    https://trustedcomputinggroup.org/wp-content/uploads/EFI-Protocol-Specification-rev13-160330final.pdf.
    Definition:
    typedef struct tdTCG_EfiSpecIdEventStruct {
        BYTE[16] signature;
        UINT32 platformClass;
        UINT8 specVersionMinor;
        UINT8 specVersionMajor;
        UINT8 specErrata;
        UINT8 uintnSize;
        UINT32 numberOfAlgorithms;
        TCG_EfiSpecIdEventAlgorithmSize[numberOfAlgorithms] digestSizes;
        UINT8 vendorInfoSize;
        BYTE[VendorInfoSize] vendorInfo;
    } TCG_EfiSpecIDEventStruct;
*/
#[derive(Clone)]
pub struct TcgEfiSpecIdEvent {
    pub signature: [u8; 16],
    pub platform_class: u32,
    pub spec_version_minor: u8,
    pub spec_version_major: u8,
    pub spec_errata: u8,
    pub uintn_ize: u8,
    pub number_of_algorithms: u32,
    pub digest_sizes: Vec<TcgEfiSpecIdEventAlgorithmSize>,
    pub vendor_info_size: u8,
    pub vendor_info: Vec<u8>,
}

impl Default for TcgEfiSpecIdEvent {
    fn default() -> Self {
        Self::new()
    }
}

impl TcgEfiSpecIdEvent {
    pub fn new() -> TcgEfiSpecIdEvent {
        TcgEfiSpecIdEvent {
            signature: [0; 16],
            platform_class: 0,
            spec_version_minor: 0,
            spec_version_major: 0,
            spec_errata: 0,
            uintn_ize: 0,
            number_of_algorithms: 0,
            digest_sizes: Vec::new(),
            vendor_info_size: 0,
            vendor_info: Vec::new(),
        }
    }
}

/***
    TCG TCG_EfiSpecIdEventAlgorithmSize defined at
    https://trustedcomputinggroup.org/wp-content/uploads/EFI-Protocol-Specification-rev13-160330final.pdf.
    Definiton:
    typedef struct tdTCG_EfiSpecIdEventAlgorithmSize {
        UINT16 algorithmId;
        UINT16 digestSize;
    } TCG_EfiSpecIdEventAlgorithmSize;
*/
#[derive(Clone)]
pub struct TcgEfiSpecIdEventAlgorithmSize {
    pub algo_id: u16,
    pub digest_size: u32,
}

#[derive(Clone)]
pub enum EventLogEntry {
    TcgImrEvent(TcgImrEvent),
    TcgPcClientImrEvent(TcgPcClientImrEvent),
}

impl EventLogEntry {
    pub fn show(&self) {
        match self {
            EventLogEntry::TcgImrEvent(tcg_imr_event) => &tcg_imr_event.show(),
            EventLogEntry::TcgPcClientImrEvent(tcg_pc_client_imr_event) => {
                &tcg_pc_client_imr_event.show()
            }
        };
    }
}

impl TcgImrEvent {
    pub fn show(&self) {
        info!(
            "        -------------------------------Event Log Entry-----------------------------"
        );
        info!("        IMR               : {}", self.imr_index);
        info!(
            "        Type              : {:02X?} ({})",
            self.event_type,
            &TcgEventType::get_event_type_string(self.event_type)
        );

        for digest_index in 0..self.digests.len() {
            info!(
                "        Algorithm_id[{}]   : {} {}",
                digest_index,
                self.digests[digest_index].algo_id,
                ALGO_NAME_MAP
                    .get(&self.digests[digest_index].algo_id)
                    .unwrap()
                    .to_owned()
            );
            info!("        Digest[{}]:", digest_index);
            dump_data(&self.digests[digest_index].hash);
        }
        info!("        Event:");
        dump_data(&self.event);
    }
}

impl TcgPcClientImrEvent {
    pub fn show(&self) {
        info!(
            "        --------------------Header Specification ID Event--------------------------"
        );
        info!("        IMR               : {}", self.imr_index);
        info!(
            "        Type              : {:02X?} ({})",
            self.event_type,
            &TcgEventType::get_event_type_string(self.event_type)
        );
        info!("        Digest:");
        dump_data(&self.digest.to_vec());
        info!("        Event:");
        dump_data(&self.event);
    }
}

File: common/golang/cctrusted_base/tcg.go
package cctrusted_base

type TcgEventFormat string
type TcgEventType uint32
type TCG_ALG int32

const (
	TPM_ALG_ERROR  TCG_ALG = 0x0
	TPM_ALG_RSA    TCG_ALG = 0x1
	TPM_ALG_SHA1   TCG_ALG = 0x4
	TPM_ALG_SHA256 TCG_ALG = 0xB
	TPM_ALG_SHA384 TCG_ALG = 0xC
	TPM_ALG_SHA512 TCG_ALG = 0xD
	TPM_ALG_ECDSA  TCG_ALG = 0x18
)

func GetDefaultTPMAlg() TCG_ALG {
	return TPM_ALG_SHA384
}

func (alg TCG_ALG) String() string {
	switch alg {
	case TPM_ALG_ERROR:
		return "TPM_ALG_ERROR"
	case TPM_ALG_RSA:
		return "TPM_ALG_RSA"
	case TPM_ALG_SHA1:
		return "TPM_ALG_SHA1"
	case TPM_ALG_SHA256:
		return "TPM_ALG_SHA256"
	case TPM_ALG_SHA384:
		return "TPM_ALG_SHA384"
	case TPM_ALG_SHA512:
		return "TPM_ALG_SHA512"
	case TPM_ALG_ECDSA:
		return "TPM_ALG_ECDSA"
	}
	return ""
}

var (
	TPM_ALG_HASH_DIGEST_SIZE_TABLE = map[TCG_ALG]int{
		TPM_ALG_SHA1:   20,
		TPM_ALG_SHA256: 32,
		TPM_ALG_SHA384: 48,
		TPM_ALG_SHA512: 64,
	}
)

const (
	TCG_PCCLIENT_FORMAT  TcgEventFormat = "tcg_pcclient"
	TCG_CANONICAL_FORMAT TcgEventFormat = "tcg_canonical"

	EV_PREBOOT_CERT            TcgEventType = 0x0
	EV_POST_CODE               TcgEventType = 0x1
	EV_UNUSED                  TcgEventType = 0x2
	EV_NO_ACTION               TcgEventType = 0x3
	EV_SEPARATOR               TcgEventType = 0x4
	EV_ACTION                  TcgEventType = 0x5
	EV_EVENT_TAG               TcgEventType = 0x6
	EV_S_CRTM_CONTENTS         TcgEventType = 0x7
	EV_S_CRTM_VERSION          TcgEventType = 0x8
	EV_CPU_MICROCODE           TcgEventType = 0x9
	EV_PLATFORM_CONFIG_FLAGS   TcgEventType = 0xa
	EV_TABLE_OF_DEVICES        TcgEventType = 0xb
	EV_COMPACT_HASH            TcgEventType = 0xc
	EV_IPL                     TcgEventType = 0xd
	EV_IPL_PARTITION_DATA      TcgEventType = 0xe
	EV_NONHOST_CODE            TcgEventType = 0xf
	EV_NONHOST_CONFIG          TcgEventType = 0x10
	EV_NONHOST_INFO            TcgEventType = 0x11
	EV_OMIT_BOOT_DEVICE_EVENTS TcgEventType = 0x12
	EV_POST_CODE2              TcgEventType = 0x13
	// IMA event type defined aligned with MSFT
	IMA_MEASUREMENT_EVENT            TcgEventType = 0x14
	EV_EFI_EVENT_BASE                TcgEventType = 0x80000000
	EV_EFI_VARIABLE_DRIVER_CONFIG    TcgEventType = EV_EFI_EVENT_BASE + 0x1
	EV_EFI_VARIABLE_BOOT             TcgEventType = EV_EFI_EVENT_BASE + 0x2
	EV_EFI_BOOT_SERVICES_APPLICATION TcgEventType = EV_EFI_EVENT_BASE + 0x3
	EV_EFI_BOOT_SERVICES_DRIVER      TcgEventType = EV_EFI_EVENT_BASE + 0x4
	EV_EFI_RUNTIME_SERVICES_DRIVER   TcgEventType = EV_EFI_EVENT_BASE + 0x5
	EV_EFI_GPT_EVENT                 TcgEventType = EV_EFI_EVENT_BASE + 0x6
	EV_EFI_ACTION                    TcgEventType = EV_EFI_EVENT_BASE + 0x7
	EV_EFI_PLATFORM_FIRMWARE_BLOB    TcgEventType = EV_EFI_EVENT_BASE + 0x8
	EV_EFI_HANDOFF_TABLES            TcgEventType = EV_EFI_EVENT_BASE + 0x9
	EV_EFI_PLATFORM_FIRMWARE_BLOB2   TcgEventType = EV_EFI_EVENT_BASE + 0xa
	EV_EFI_HANDOFF_TABLES2           TcgEventType = EV_EFI_EVENT_BASE + 0xb
	EV_EFI_VARIABLE_BOOT2            TcgEventType = EV_EFI_EVENT_BASE + 0xc
	EV_EFI_GPT_EVENT2                TcgEventType = EV_EFI_EVENT_BASE + 0xd
	EV_EFI_HCRTM_EVENT               TcgEventType = EV_EFI_EVENT_BASE + 0x10
	EV_EFI_VARIABLE_AUTHORITY        TcgEventType = EV_EFI_EVENT_BASE + 0xe0
	EV_EFI_SPDM_FIRMWARE_BLOB        TcgEventType = EV_EFI_EVENT_BASE + 0xe1
	EV_EFI_SPDM_FIRMWARE_CONFIG      TcgEventType = EV_EFI_EVENT_BASE + 0xe2
	EV_EFI_SPDM_DEVICE_POLICY        TcgEventType = EV_EFI_EVENT_BASE + 0xe3
	EV_EFI_SPDM_DEVICE_AUTHORITY     TcgEventType = EV_EFI_EVENT_BASE + 0xe4
)

func (t TcgEventType) String() string {
	switch t {
	case EV_PREBOOT_CERT:
		return "EV_PREBOOT_CERT"
	case EV_POST_CODE:
		return "EV_POST_CODE"
	case EV_UNUSED:
		return "EV_UNUSED"
	case EV_NO_ACTION:
		return "EV_NO_ACTION"
	case EV_SEPARATOR:
		return "EV_SEPARATOR"
	case EV_ACTION:
		return "EV_ACTION"
	case EV_EVENT_TAG:
		return "EV_EVENT_TAG"
	case EV_S_CRTM_CONTENTS:
		return "EV_S_CRTM_CONTENTS"
	case EV_S_CRTM_VERSION:
		return "EV_S_CRTM_VERSION"
	case EV_CPU_MICROCODE:
		return "EV_CPU_MICROCODE"
	case EV_PLATFORM_CONFIG_FLAGS:
		return "EV_PLATFORM_CONFIG_FLAGS"
	case EV_TABLE_OF_DEVICES:
		return "EV_TABLE_OF_DEVICES"
	case EV_COMPACT_HASH:
		return "EV_COMPACT_HASH"
	case EV_IPL:
		return "EV_IPL"
	case EV_IPL_PARTITION_DATA:
		return "EV_IPL_PARTITION_DATA"
	case EV_NONHOST_CODE:
		return "EV_NONHOST_CODE"
	case EV_NONHOST_CONFIG:
		return "EV_NONHOST_CONFIG"
	case EV_NONHOST_INFO:
		return "EV_NONHOST_INFO"
	case EV_OMIT_BOOT_DEVICE_EVENTS:
		return "EV_OMIT_BOOT_DEVICE_EVENTS"
	case EV_POST_CODE2:
		return "EV_POST_CODE2"
	case IMA_MEASUREMENT_EVENT:
		return "IMA_MEASUREMENT_EVENT"
	case EV_EFI_EVENT_BASE:
		return "EV_EFI_EVENT_BASE"
	case EV_EFI_VARIABLE_DRIVER_CONFIG:
		return "EV_EFI_VARIABLE_DRIVER_CONFIG"
	case EV_EFI_VARIABLE_BOOT:
		return "EV_EFI_VARIABLE_BOOT"
	case EV_EFI_BOOT_SERVICES_APPLICATION:
		return "EV_EFI_BOOT_SERVICES_APPLICATION"
	case EV_EFI_BOOT_SERVICES_DRIVER:
		return "EV_EFI_BOOT_SERVICES_DRIVER"
	case EV_EFI_RUNTIME_SERVICES_DRIVER:
		return "EV_EFI_RUNTIME_SERVICES_DRIVER"
	case EV_EFI_GPT_EVENT:
		return "EV_EFI_GPT_EVENT"
	case EV_EFI_ACTION:
		return "EV_EFI_ACTION"
	case EV_EFI_PLATFORM_FIRMWARE_BLOB:
		return "EV_EFI_PLATFORM_FIRMWARE_BLOB"
	case EV_EFI_HANDOFF_TABLES:
		return "EV_EFI_HANDOFF_TABLES"
	case EV_EFI_PLATFORM_FIRMWARE_BLOB2:
		return "EV_EFI_PLATFORM_FIRMWARE_BLOB2"
	case EV_EFI_HANDOFF_TABLES2:
		return "EV_EFI_HANDOFF_TABLES2"
	case EV_EFI_VARIABLE_BOOT2:
		return "EV_EFI_VARIABLE_BOOT2"
	case EV_EFI_GPT_EVENT2:
		return "EV_EFI_GPT_EVENT2"
	case EV_EFI_HCRTM_EVENT:
		return "EV_EFI_HCRTM_EVENT"
	case EV_EFI_VARIABLE_AUTHORITY:
		return "EV_EFI_VARIABLE_AUTHORITY"
	case EV_EFI_SPDM_FIRMWARE_BLOB:
		return "EV_EFI_SPDM_FIRMWARE_BLOB"
	case EV_EFI_SPDM_FIRMWARE_CONFIG:
		return "EV_EFI_SPDM_FIRMWARE_CONFIG"
	case EV_EFI_SPDM_DEVICE_POLICY:
		return "EV_EFI_SPDM_DEVICE_POLICY"
	case EV_EFI_SPDM_DEVICE_AUTHORITY:
		return "EV_EFI_SPDM_DEVICE_AUTHORITY"
	}
	return ""
}
